Add support for remote tier management

krisis opened the pull request:
## Description
With this change MinIO's ILM supports transitioning objects to a remote tier.
This change includes support for Azure Blob Storage, AWS S3 and Google Cloud
Storage as remote tier storage backends.

Co-authored-by: Poorna Krishnamoorthy <poorna@minio.io>
Co-authored-by: Krishna Srinivas <krishna@minio.io>

## Motivation and Context


## How to test this PR?


## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

krisis committed:
Add support for remote tier management

With this change MinIO's ILM supports transitioning objects to a remote tier.
This change includes support for Azure Blob Storage, AWS S3 compatible object
storage incl. MinIO and Google Cloud Storage as remote tier storage backends.

Some new additions include,
 - Admin APIs remote tier configuration management

 - Simple journal to track remote objects to be 'collected'
   This is used by object API handlers which 'mutate' object versions by
   overwriting/replacing content (Put/CopyObject) or removing the version
   itself (e.g DeleteObjectVersion).

 - Rework of previous ILM transition to fit the new model
   In the new model, a storage-class (a.k.a remote tier) is defined by the
   'remote' object storage type (one of s3, azure, gcs), bucket name and a
   prefix.

Co-authored-by: Poorna Krishnamoorthy <poorna@minio.io>
Co-authored-by: Krishna Srinivas <krishna@minio.io>

vadmeste commented on cmd/endpoint.go:
```diff
@@ -256,6 +256,20 @@ func (l EndpointServerPools) Localhost() string {
+// FirstLocalDiskPath returns the disk path of first (in cmdline args order)
+// local endpoint. Returns empty string and false if no local endpoint was
+// found.
+func (l EndpointServerPools) FirstLocalDiskPath() (string, bool) {
```
We don't need to return bool here, we always have a first local disk path

vadmeste commented on cmd/http/headers.go:
```diff
@@ -167,6 +167,8 @@ const (
 	MinIOSourceProxyRequest = "X-Minio-Source-Proxy-Request"
 	// Header indicates that this request is a replication request to create a REPLICA
 	MinIOSourceReplicationRequest = "X-Minio-Source-Replication-Request"
+	// Object date/time of expiration
```
You mean transition here ?

vadmeste commented on cmd/peer-rest-common.go:
```diff
@@ -24,41 +24,42 @@ const (
+	peerRESTMethodGetMetacacheListing      = "/getmetacache"
+	peerRESTMethodUpdateMetacacheListing   = "/updatemetacache"
+	peerRESTMethodGetPeerMetrics           = "/peermetrics"
+	peerRESTMethodLoadTransitionTierConfig = "/loadtransitiontierconfig"
```
you need to update peerRESTVersion

vadmeste commented on cmd/tier-journal.go:
```diff
@@ -0,0 +1,277 @@
+		diskPath: diskPath,
+	}
+
+	if err := os.MkdirAll(filepath.Dir(j.JournalPath()), os.FileMode(0770)); err != nil {
```
I guess you want 0600

vadmeste commented on pkg/madmin/tier.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+// ListTiers returns a list of remote tiers configured.
+func (adm *AdminClient) ListTiers(ctx context.Context) ([]*TierConfig, error) {
```
Let's encrypt the content of this as well (peace of mind even if it is not necessary)

vadmeste commented on cmd/erasure-server-pool.go:
```diff
@@ -1754,3 +1754,33 @@ func (z *erasureServerPools) GetObjectTags(ctx context.Context, bucket, object s
+		return z.serverPools[0].TransitionObject(ctx, bucket, object, opts)
+	}
+	// We don't know the size here set 1GiB atleast.
+	idx, err := z.getPoolIdx(ctx, bucket, object, 1<<30)
```
We need to make sure that the object exists right ? getPooldx() does not return an error in that case

poornas replied:
if object does not exist, the GetObjectInfo done in TransitionObject should return err

vadmeste commented on cmd/erasure-server-pool.go:
```diff
@@ -1754,3 +1754,33 @@ func (z *erasureServerPools) GetObjectTags(ctx context.Context, bucket, object s
+		return z.serverPools[0].RestoreTransitionedObject(ctx, bucket, object, opts)
+	}
+	// We don't know the size here set 1GiB atleast.
+	idx, err := z.getPoolIdx(ctx, bucket, object, 1<<30)
```
We need to make sure that the object exists right ? getPooldx() does not return an error in that case

poornas replied:
if object does not exist, the GetObjectInfo done in RestoreTransitionedObject should return err

vadmeste commented on cmd/tier.go:
```diff
@@ -0,0 +1,351 @@
+	if err != nil {
+		return nil, nil, err
+	}
+	if GlobalKMS == nil {
```
Do we use KMS to store MinIO config ? maybe this is an overkill, not sure but I think MinIO config is encrypted with root secret key

vadmeste commented on cmd/tier.go:
```diff
@@ -0,0 +1,351 @@
+	}
+}
+
+func loadTransitionTierConfig() (*TierConfigMgr, error) {
```
You might need a retry for this function (not sure) otherwise it could be orchestrated in initSubSystem()s

vadmeste commented on cmd/warm-backend-gcs.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+func (gcs *warmBackendGCS) Remove(ctx context.Context, key string) error {
+	err := gcs.client.Bucket(gcs.Bucket).Object(key).Delete(ctx)
```
I think you forget to add prefix here

vadmeste commented on cmd/warm-backend-s3.go:
```diff
@@ -0,0 +1,127 @@
+func (s3 *warmBackendS3) Get(ctx context.Context, object string, opts warmBackendGetOpts) (io.ReadCloser, error) {
+	gopts := minio.GetObjectOptions{}
+
+	if opts.startOffset >= 0 && opts.length >= 0 {
```
You mean `||` not `&&` I think

poornas replied:
offset and length has already been calculated by NewGetObjectReader, this is &&

vadmeste replied:
> offset and length has already been calculated by NewGetObjectReader, this is &&

Okay.

What  I suggest is we void this use case, startOffset = 0 and length = 0, because it will lead gopts.SetRange(0, -1)

poornas replied:
accounted for startOffset = 0 and length = 0 in the above condition

vadmeste commented on cmd/warm-backend-s3.go:
```diff
@@ -0,0 +1,127 @@
+	if err != nil {
+		return false, s3.ToObjectError(err)
+	}
+	if len(result.CommonPrefixes) > 0 {
```
Why you check only for common prefixes ? better to make sure that the prefix is empty, so it is not used for other uses

poornas replied:
changing this to below should cover that case, wouldn't it?

```suggestion
	if len(result.CommonPrefixes) >= 0 {
```

vadmeste replied:
> changing this to below should cover that case, wouldn't it?

so I think that we want tier prefix to be completely empty, in that case we want to make sure that `result.CommonPrefix == 0` and `len(result.Objects) == 0` as well

krisis replied:
@vadmeste understood. The condition in this PR checks if the given prefix is being used by another ILM tier but doesn't account for it being used by other applications. We should add `len(result.Objects) == 0`.

vadmeste commented on cmd/tier-sweeper.go:
```diff
@@ -0,0 +1,147 @@
+//     err := globalTierJournal.AddEntry(jentry)
+//     logger.LogIf(ctx, err)
+// }
+type objSweeper struct {
```
maybe tierObject is a better name, it doesn't sweep anything

vadmeste commented on cmd/object-handlers.go:
```diff
@@ -1320,6 +1320,21 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 		objInfo.ETag = remoteObjInfo.ETag
 		objInfo.ModTime = remoteObjInfo.LastModified
 	} else {
+		var hasLifecycleConfig bool
```
I think it is possible to simplify the code here, let's do for example:

ilmTask := createILMTask(bucket, object, get_info_func), it can be almost no-op if lifecycle is not enabled

then call this later

ilmTask.removeInTierIfNecessary()

vadmeste reviewed (commented):
Some comments, not tested yet.. probably need a second a review

poornas commented on cmd/data-scanner.go:
```diff
@@ -955,16 +957,19 @@ func evalActionFromLifecycle(ctx context.Context, lc lifecycle.Lifecycle, obj Ob
+		// mark transition as pending
+		obj.UserDefined[ReservedMetadataPrefixLower+TransitionStatus] = lifecycle.TransitionPending
+		obj.metadataOnly = true // Perform only metadata updates.
+		// Ignore errMethodNotAllowed error. It indicates that the object version scanned is a delete marker.
```
instead of trying to transition a delete marker,  return false if obj.DeleteMarker is true

poornas commented on cmd/object-api-errors.go:
```diff
@@ -666,6 +673,12 @@ func isErrVersionNotFound(err error) bool {
+func isErrSignatureDoesNotMatch(err error) bool {
+	var signatureDoesNotMatch SignatureDoesNotMatch
+	return errors.As(err, &signatureDoesNotMatch)
+}
```
seems like left over from rebase conflicts - master no longer has this function

krisis replied:
This is function is used to identify a signature mismatch error and translate it to "invalid tier credentials".

poornas commented on cmd/xl-storage.go:
```diff
@@ -890,6 +890,13 @@ func (s *xlStorage) DeleteVersion(ctx context.Context, volume, path string, fi F
 	}
 
+	var transitionStatus string
+	xfi, err := xlMeta.ToFileInfo(volume, path, fi.VersionID)
```
@kp, this is the cause of the `mc rm` failure - fi.VersionID would come in with a random uuid to create a "Deleted" version entry if there is a rm operation. This check has to be thought out better

klauspost commented on cmd/bucket-lifecycle.go:
```diff
@@ -150,8 +154,9 @@ func (t *transitionState) addWorker(ctx context.Context, objectAPI ObjectLayer)
+
 				if err := transitionObject(ctx, objectAPI, oi); err != nil {
-					logger.LogIf(ctx, err)
+					logger.LogIf(ctx, fmt.Errorf("Transition failed for %s/%s version:%s with %s", oi.Bucket, oi.Name, oi.VersionID, err))
```
Wrap the error:
```suggestion
					logger.LogIf(ctx, fmt.Errorf("Transition failed for %s/%s version:%s with %w", oi.Bucket, oi.Name, oi.VersionID, err))
```

klauspost commented on cmd/bucket-lifecycle.go:
```diff
@@ -172,308 +177,159 @@ func initBackgroundTransition(ctx context.Context, objectAPI ObjectLayer) {
+	var opts ObjectOptions
+	opts.Versioned = globalBucketVersioningSys.Enabled(bucket)
+	opts.VersionID = lcOpts.VersionID
+	switch action {
```
Check for unknown action values, add `default:` And maybe an empty value for `actionNone`.

klauspost commented on cmd/common-main.go:
```diff
@@ -342,6 +342,12 @@ func handleCommonEnvVars() {
 	}
+	if tiers := env.Get("_MINIO_DEBUG_REMOTE_TIERS_IMMEDIATELY", ""); tiers != "" {
+		globalDebugRemoteTiersImmediately = strings.Split(tiers, ",")
+		if len(globalDebugRemoteTiersImmediately) == 0 {
```
This can never be true. Any non-empty string will produce at len() > 0 output.

klauspost commented on cmd/tier-journal.go:
```diff
@@ -0,0 +1,273 @@
+)
+
+func initTierDeletionJournal(done <-chan struct{}) (*tierJournal, error) {
+	diskPath := globalEndpoints.FirstLocalDiskPath()
```
Not a huge fan of relying on a single disk for this kind of operation.

klauspost commented on cmd/erasure-object.go:
```diff
@@ -1299,3 +1295,124 @@ func (er erasureObjects) GetObjectTags(ctx context.Context, bucket, object strin
+	}
+	defer lk.Unlock()
+
+	fi, metaArr, onlineDisks, err := er.getObjectFileInfo(ctx, bucket, object, opts, false)
```
You are likely to need the data.
```suggestion
	fi, metaArr, onlineDisks, err := er.getObjectFileInfo(ctx, bucket, object, opts, true)
```

klauspost replied:
(it will be required for #11758 anyway)

klauspost commented on cmd/erasure-server-pool.go:
```diff
@@ -1773,3 +1773,33 @@ func (z *erasureServerPools) GetObjectTags(ctx context.Context, bucket, object s
+		return z.serverPools[0].TransitionObject(ctx, bucket, object, opts)
+	}
+	// We don't know the size here set 1GiB atleast.
+	idx, err := z.getPoolIdx(ctx, bucket, object, 1<<30)
```
Aren't we transitioning FROM this `z`? If so, we don't need free space, and we just need the index of the existing object.
```suggestion
	idx, err := z.getPoolIdx(ctx, bucket, object, 0)
```
Added https://github.com/minio/minio/pull/11799 which is even more appropriate.

krisis replied:
Thanks @klauspost for the PR. I have rebased to use getPoolIdxExisting.

klauspost commented on cmd/object-api-datatypes.go:
```diff
@@ -164,7 +164,9 @@ type ObjectInfo struct {
-	TransitionStatus string
+	TransitionStatus    string
+	transitionedObjName string // name of transitioned object on remote target
+	TransitionTier      string // name of MinIO defined storage class object transitioned to.
```
nit: Doc above as other vars.

klauspost commented on cmd/object-api-datatypes.go:
```diff
@@ -164,7 +164,9 @@ type ObjectInfo struct {
 	// TransitionStatus indicates if transition is complete/pending
-	TransitionStatus string
+	TransitionStatus    string
+	transitionedObjName string // name of transitioned object on remote target
```
Why is this not exported? Do we need it internally?

klauspost commented on cmd/tier-journal.go:
```diff
@@ -0,0 +1,273 @@
+	}
+	if n < 2 {
+		return nil, errors.New("invalid pending deletes journal format")
+	}
```
```suggestion
	var data [tierJournalHdrLen]byte	
	if err := io.ReadFull(file, data[:]); err != nil {
		return nil, err
	}
```

klauspost commented on cmd/tier-sweeper.go:
```diff
@@ -0,0 +1,155 @@
+
+// versionID interface is used to fetch object versionID from disparate sources
+// like http.Request and ObjectToDelete.
+type versionID interface {
```
```suggestion
type versionIDer interface {
```

klauspost commented on cmd/tier-sweeper.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+// ForDelete sets the version ID from v
+func (os *objSweeper) ForDelete(v versionID) *objSweeper {
```
`ForDelete` doesn't tell much. Maybe:

```suggestion
func (os *objSweeper) DeleteVersion(v versionID) *objSweeper {
```
or if it makes sense:

```suggestion
func (os *objSweeper) WithVersion(v versionID) *objSweeper {
```

klauspost commented on cmd/tier.go:
```diff
@@ -0,0 +1,376 @@
+}
+
+// Add adds tier to config if it passes all validations.
+func (config *TierConfigMgr) Add(tier madmin.TierConfig) error {
```
```suggestion
func (config *TierConfigMgr) Add(ctx context.context, tier madmin.TierConfig) error {
```

klauspost commented on cmd/warm-backend.go:
```diff
@@ -0,0 +1,79 @@
+
+// checkWarmBackend checks if tier config credentials have sufficient privileges
+// to perform all operations definde in the WarmBackend interface.
+// FIXME: currently, we check only for Get.
+func checkWarmBackend(ctx context.Context, w WarmBackend) error {
+	// TODO: requires additional checks to ensure that WarmBackend
+	// configuration has sufficient privileges to Put/Remove objects as well.
```
Will these be added? Seems like a good idea.

krisis replied:
These checks will be added in a subsequent PR.

klauspost commented on pkg/madmin/tier-config_test.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+// TestUnmarshalInvalidTierConfig tests that TierConfig parsing can catch invalid tier configs
+func TestUnmarshalInvalidTierConfig(t *testing.T) {
```
Now that we are talking about tests, this is not really that covering ;O

klauspost reviewed (commented):
Nothing major. Great work.

Could do with a lot more tests, though.

krisis committed:
Fixed bugs, review comments and more unit-tests

- Leverage inline small object feature
- Migrate legacy objects to latest object format before transitioning
- Fix restore to particular version if specified
- Extend SharedDataDirCount to handle transitioned and restored objects

harshavardhana commented on cmd/xl-storage-format-v2.go:
```diff
@@ -1011,6 +1011,55 @@ func (j xlMetaV2Object) ToFileInfo(volume, path string) (FileInfo, error) {
+	}
+	if given == nil {
+		return 0
+	}
```
You can write it like 

```
        var given *xlMetaV2Object
	switch v.Type {
        default: 
               return 0
	case ObjectType:
		given = v.ObjectV2
	}
```

harshavardhana commented on cmd/xl-storage-format-v2.go:
```diff
@@ -1011,6 +1011,55 @@ func (j xlMetaV2Object) ToFileInfo(volume, path string) (FileInfo, error) {
+		return uuid.UUID(barr).String()
+	}
+	// skip if v's data is inlined
+	if z.data.find(toUUIDStr(given.VersionID)) != nil {
```
```
if z.data.find(toUUIDStr(given.VersionID)) != nil || z.data.find(toUUIDStr(given.DataDir)) != nil {
```

harshavardhana commented on cmd/xl-storage-format-v2.go:
```diff
@@ -1011,6 +1011,55 @@ func (j xlMetaV2Object) ToFileInfo(volume, path string) (FileInfo, error) {
+			}
+			if z.data.find(toUUIDStr(u.DataDir)) != nil {
+				continue
+			}
```
Glad you found this bug!

harshavardhana reviewed (changes requested):
Actually you can have the same findDataDir loop to find SharedDataDirCount() - you don't need a top level check like `z.Type` @krisis 

Also you should only send the dataDir that the potential to have DataDir there is no point making this function generic. We are only interested in DataDir when the ObjectV2 is found so passing 

SharedDataDirCount(version.ObjectV2) is enough - makes your code super simple and not very generic.
